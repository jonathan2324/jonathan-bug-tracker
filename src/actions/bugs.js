import uuid from 'uuid'

//ADD_BUG
export const addBug = (
    { 
        name = '',
        description = '',
        notes = '',
        severity = 'low',
        completed = false,
        createdAt = 0  } = {}
    ) => ({
    type: 'ADD_BUG',
    bug: {
        id: uuid(),
        name,
        description,
        notes,
        severity,
        completed,
        createdAt
    }
    
})

//REMOVE_BUG

export const removeBug = ({ id }) => ({
    type: 'REMOVE_BUG',
    id
})

//Edit Bug
export const editBug = (id, updates) => ({
    type: 'EDIT_BUG',
    id,
    updates
})

//TOGGLE COMPLETED
export const toggleCompletedBug = (id, completed) => ({
    type: 'TOGGLE_COMPLETED_BUG',
    id,
    completed
})