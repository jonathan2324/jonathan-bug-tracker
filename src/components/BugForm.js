import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'



const now = moment()
console.log(now.format('MMM Do YYYY'))

class BugForm extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            name: props.bug ? props.bug.name : '',
            severity: props.bug ? props.bug.severity : 'low',
            description: props.bug ? props.bug.description : '',
            notes: props.bug ? props.bug.notes : '',
            createdAt: props.bug ? moment(props.bug.createdAt) : moment(),
            calendarFocused: false,
            error: ''
    
        }
    }
    

    onNameChange = (e) => {
        const name = e.target.value 
        this.setState(() => ({ name }))
    }

    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }

    onNotesChange = (e) => {
        const notes = e.target.value
        this.setState(() => ({ notes }))       
    }

    onSeverityChange = (e) => {
        const severity = e.target.value
        this.setState(() => ({ severity }))
    }

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({
                createdAt
            }))

        }

    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }

    onSubmit = (e) => {
        e.preventDefault()

        if (!this.state.name || !this.state.description) {
            this.setState(() => ({ error: 'Please provide an employee name and a description'}))
        } else {
            this.setState(() => ({ error: '' }))
            this.props.onSubmit({
                description: this.state.description,
                notes: this.state.notes,
                name: this.state.name,
                createdAt: this.state.createdAt.valueOf(),
                severity: this.state.severity
            })
        }
    }

    render () {
        return (
            <div>
                <p>{this.state.error}</p>
                <form onSubmit={this.onSubmit}>
                    <input type='text' placeholder='Employee name' autoFocus value={this.state.name} onChange={this.onNameChange}/>
                    <select value={this.state.severity} onChange={this.onSeverityChange}>
                        <option value="urgent">Urgent</option>
                        <option value="moderate">Moderate</option>
                        <option value="low">Low</option>
                    </select>

                    <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    />

                    <textarea type='text' maxLength="50" placeholder='Enter error description (max length 50 characters)' value={this.state.description} onChange={this.onDescriptionChange}/>
                    <textarea type='text' placeholder='Enter important notes about error' value={this.state.notes} onChange={this.onNotesChange}/>
                    <button>Add Bug</button>

                </form>
            </div>
        )
    }
}

export default BugForm