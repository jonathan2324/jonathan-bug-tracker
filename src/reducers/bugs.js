

//bugReducer
const bugReducerDefaultState = []

const bugReducer = (state = bugReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_BUG':
            return [
                ...state,
                action.bug
            ]
        case 'REMOVE_BUG':
            return state.filter(({ id }) => id !== action.id)
        case 'EDIT_BUG':
            return state.map((bug) => {
                if (bug.id === action.id) {
                    return {
                        ...bug,
                        ...action.updates
                    }
                } else {
                    return bug
                }

            })
        case 'TOGGLE_COMPLETED_BUG':
            return state.map((bug) => {
                if (bug.id === action.id) {
                    return {
                        ...bug,
                        completed: !action.completed
                    }
                } else {
                    return bug
                }
            })
        case 'ADD_CONTRIBUTION':
            return state.map((bug) => {
                if(bug.id === action.id){
                    return {
                        ...bug,
                        contributions: [
                            
                            action.contribution,
                            ...bug.contributions,
                            
                        ]
                    }
                } else return bug
            })
        default:
            return state;
    }
};

export default bugReducer