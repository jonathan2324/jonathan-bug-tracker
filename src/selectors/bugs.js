import moment from 'moment'

//Get visible bugs
const getVisibleBugs = (bugs, { text, sortBy, startDate, endDate }) => {
    return bugs.filter((bug) => {
        const createdAtMoment = moment(bug.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day'): true
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
        const textMatch = bug.name.toLowerCase().includes(text.toLowerCase())
        
        return startDateMatch && endDateMatch && textMatch
    }).sort((a, b) => {
        if(sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        }
        else if (sortBy === 'severity') {
            if (a.severity === 'urgent') {
                return a.severity < b.severity ? 1 : -1
            } else if (a.severity === 'moderate') {
                return a.severity > b.severity ? -1 : 1 
            } else {
                return 0
            }
        }
        else {
            return null
        }
    }).sort((a,b) => {
       if(sortBy === 'completed') {
           return a.completed ? 1 : -1
       } else {
           return 0
       }
    })
}

export default getVisibleBugs