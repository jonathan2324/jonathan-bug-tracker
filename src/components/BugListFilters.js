import React from 'react'
import {connect} from 'react-redux'
import { setTextFilter, sortByDate, sortBySeverity, setStartDate, setEndDate, sortByCompleted } from '../actions/filters'
import { DateRangePicker } from 'react-dates'



export class BugListFilters extends React.Component {
    state = {
        calendarFocused: null
    }

    onDatesChange = ({startDate, endDate}) => { //destructure object that is passed into this function by the date range picker
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)        //these are the startDate and endDate that was required by airbnb
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused: calendarFocused }))
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
        }
    
    onSortChange = (e) => {
        if (e.target.value === "date") {
            this.props.sortByDate()
        } else if (e.target.value === 'severity') {
            this.props.sortBySeverity()
        } else if (e.target.value === 'completed') {
            this.props.sortByCompleted()
        }
    }

    render() {
        return (
            <div>
                <input type='text' value={this.props.filters.text} onChange={this.onTextChange}/>
                <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                    <option value='date'>Date</option>
                    <option value='severity'>Severity</option>
                    <option value='completed'>Completed status</option>
                </select>
                <DateRangePicker 
                startDate={this.props.filters.startDate} 
                endDate={this.props.filters.endDate}
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.calendarFocused}
                onFocusChange={this.onFocusChange}
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange={() => false}
                />
            </div>
        )
    }

}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortBySeverity: () => dispatch(sortBySeverity()),
    sortByCompleted: () => dispatch(sortByCompleted()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(BugListFilters)