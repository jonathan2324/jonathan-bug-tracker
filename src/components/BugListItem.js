import React from 'react'
import { Link } from 'react-router-dom'
import {connect} from 'react-redux'
import { toggleCompletedBug } from '../actions/bugs'

export const BugListItem = ({ dispatch, id, name, description, severity, createdAt, completed}) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>
                {name}
            </h3>
            <p>{description}</p>
            <p>
                {severity} - {createdAt}
            </p>
        </Link>

        <button onClick={(e) => dispatch(toggleCompletedBug(id, completed))}>{completed ? 'Closed': 'Open'}</button>
    </div>


)


export default connect()(BugListItem)