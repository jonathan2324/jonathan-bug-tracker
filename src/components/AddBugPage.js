import React from 'react';
import BugForm from './BugForm'
import { connect } from 'react-redux'
import { addBug } from '../actions/bugs'

export class AddBugPage extends React.Component {
  onSubmit = (bug) => {
    this.props.addBug(bug)
    this.props.history.push('/')
  }
  render() {
    return (
      <div>
        <h1>Add bug</h1>
        <BugForm onSubmit={this.onSubmit}/>
      </div>
    )
  }
}


const mapDispatchToProps = (dispatch) => {
  return {
    addBug:(bug) => dispatch(addBug(bug))
  }
}

export default connect(undefined, mapDispatchToProps)(AddBugPage);