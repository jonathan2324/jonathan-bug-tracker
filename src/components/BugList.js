import React from 'react'
import { connect } from 'react-redux'
import BugListItem from './BugListItem'
import selectBugs from '../selectors/bugs'

export const BugList = (props) => (
    <div>

        {
            props.bugs.length === 0 ? (

                <p>No bugs</p>

            ) : (

                props.bugs.map((bug) => {
                    return <BugListItem key={bug.id} {...bug}/>
                 })

            )
        }



    </div>
)

const mapStateToProps = (state) => {
    return {
        bugs: selectBugs(state.bugs, state.filters)
    }
}

const ConnectedBugList = connect(mapStateToProps)(BugList)

export default ConnectedBugList