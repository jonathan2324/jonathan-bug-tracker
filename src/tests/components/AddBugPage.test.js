import React from 'react'
import { shallow } from 'enzyme'
import { AddBugPage } from '../../components/AddBugPage'
import toJSON from 'enzyme-to-json'
import bugs from '../fixtures/bugs'

let startAddBug, history, wrapper
beforeEach(() => { //sets all these variables before each test case is ran- more efficient
    startAddBug = jest.fn() 
     history = { push: jest.fn() } 
     wrapper = shallow(<AddBugPage startAddBug={startAddBug} history={history}/>) //didnt need to provide bug because none intially

})

test('Snapshot of AddBugPage', () => {

    expect(toJSON(wrapper)).toMatchSnapshot()
})

test('Should handle onSubmit', () => {
    //const onSubmit = jest.fn() //original componenet expects onSubmit
    //const history = { push: jest.fn() } //the original componenet gets a history object
    
    wrapper.find('BugForm').prop('onSubmit')(bugs[0])
    expect(history.push).toHaveBeenLastCalledWith('/') //this is the only thing that history does
    expect(startAddBug).toHaveBeenLastCalledWith(bugs[0]) //because the original just takes an object and passes it down

})