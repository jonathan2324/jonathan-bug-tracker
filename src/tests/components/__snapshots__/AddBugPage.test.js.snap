// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot of AddBugPage 1`] = `
<div>
  <h1>
    Add bug
  </h1>
  <BugForm
    onSubmit={[Function]}
  />
</div>
`;
